### ISSUE TITLE
Build Kind signature problem in v1beta1

### ISSUE THREAD

**Issue by ba32107** [grafeas/grafeas#446]
Hi!

I am currently exploring the API, and noticed something that seems to be an inconsistency. The Build Kind has the [signature field](https://github.com/grafeas/grafeas/blob/master/proto/v1beta1/build.proto#L34) in the Note itself. I don't understand why. For one particular Note, there can be many Build Occurrences. How do we know which Occurrence the signature is for? Shouldn't the signature be a field of the [Occurrence](https://github.com/grafeas/grafeas/blob/master/proto/v1beta1/build.proto#L80)?

Did I misunderstand how this Kind works?

## Expected Behavior
The `BuildSignature` field is part of `grafeas.v1beta1.build.Details` (the Occurrence), and not `grafeas.v1beta1.build.Build` (the Note).

## Actual Behavior
The `BuildSignature` field is part of the Note:
https://github.com/grafeas/grafeas/blob/master/proto/v1beta1/build.proto#L34

## Additional info
I have also noticed that the V1 version [completely lacks](https://github.com/grafeas/grafeas/blob/master/proto/v1/build.proto) the signature for the Build Kind. Could I get some insight into why this was removed? I believe it's very useful to be able to cryptographically verify build provenances.

**Comment by ba32107** [comment]
Hi @aysylu, could you please shed some light on this? This is something that we'd like to start integrating soon in our CI system, but I don't quite understand how the Build Kind is supposed to be signed/verified. Thanks!

**Comment by wkozlik** [comment]
Hi Balazs,
Recently there was a similar question asked about Grafeas Build Notes/Occurrences on the grafeas-users group. I hope [Catherine's answer](https://groups.google.com/g/grafeas-users/c/nBOwKZ3744M/m/-rnbUA8uBAAJ) makes it a bit more clear.

**Comment by ba32107** [comment]
Hi @wkozlik,

Thanks for that link. Unfortunately, it does not help, I still don't quite understand the answer to my question. Cathrine says in her answer:
> If your builder is creating three images, you should have three occurrences (one for each) and three signatures for each occurrence.

My question is: how can I create three different signatures for each occurrence, if the `BuildSignature` field is on the Note, and not the Occurrence?

**Comment by aysylu** [comment]
> > If your builder is creating three images, you should have three occurrences (one for each) and three signatures for each occurrence.
> 
> My question is: how can I create three different signatures for each occurrence, if the `BuildSignature` field is on the Note, and not the Occurrence?

@catherinejones could you please clarify?

**Comment by ba32107** [comment]
Hi @aysylu, in order to prevent any possible misunderstanding, I'd like to highlight that my confusion is not regarding Cathrine's answer - it is regarding the documentation for `v1beta1` Build Note/Occurrence.

I would also like to understand why the `v1` version of the Build Kind has no signature fields. I think it's very useful to be able to cryptographically verify build provenances - how can we do this with `v1`?

Thanks

**Comment by aysylu** [comment]
@ba32107: yes, happy to clarify this. I'm going to answer both questions, about `v1` and about the signature on the Note (what I understand was asked earlier):

1) Why doesn't `v1` have a signature field? The intention is for the Build kind to have a signature, but we'd ideally combine it with the Signature in `common.proto` which is also used in Attestation. The reason it wasn't added right away is we wanted to leave the opportunity for the community and current users of the Build to chime in on whether the Signature in `common.proto` is a good representation going forward. It wasn't added right away because in a public proto API it's easier to add fields whereas removal/update of fields requires careful consideration of the backwards compatibility. From a few informal conversations it sounded like the `common.proto` Signature would work for the Build as well, so let us know if you have thoughts on this.

2) Note has the signature field because the intention is that the BuildProvenance in the Occurrence would be signed and the Build Note created with this signature. Afterwards, the Occurrence that's attached to this Note is created, with the BuildProvenance already computed from before and the byte representation of the provenance, to ensure the signature is verified against the up-to-date provenance. If more than 1 image has the same BuildProvenance, each of these images' Occurrences would attach to the same Build Note.

Does this answer your question?

**Comment by ba32107** [comment]
Hi @aysylu, thanks for the detailed reply.

1. Great, it seems like we are on the same page about the Build Kind having a signature. I believe the `Signature` field in `common.proto` is a good one: it is generic enough that we are able to use it with any type of signing mechanism (as opposed to `v1beta1`'s `PgpSignedAttestation`, which forces PGP). So I do believe this is a good way forward! Having Build use the `Signature` field in `v1` sounds great from our perspective.

2. I'm afraid I still don't quite understand. Perhaps I misinterpreted what the Build Note and Occurrence represents? Let me describe from my perspective (note that I am talking about `v1beta1` now):

**Build Note**: this represents a Builder instance. In my mind, this is equivalent to a CI server, e.g. Jenkins. It has a version field, which I think about as "Jenkins  2.222.4" for example. I am basing this assumption on the docs, which say "Note holding the version of the provider's builder" and "Version of the builder which produced this build." 

**Build Occurrence**: this represents a particular run of a build. It has the `BuildProvenance` field, which describes pretty much everything about a build (source repos, commands, environment variables, log file URL, build ID). So in my mind, if a Jenkins job runs three times, each will produce a different build provenance, even if they belong to the same job. They would have different build IDs, different log URLs, they would be built from different source revisions, etc. Therefore, they have different signatures.

A Jenkins server can have multiple jobs, and each job can have multiple runs, or in other words, multiple build provenances. So in Grafeas concepts, this means that each Build Note has a lot of Build Occurrences. 

According to my logic, each of those occurrences must have different signatures. So what I am not understanding is, how come that the Note has the signature field? Continuing the Jenkins analogy, this would mean that we are storing just one build provenance for a Jenkins server instance. It's as if the Jenkins server would only be able to execute one build, and nothing else.

It seems to me that I may be misunderstanding what the Build Note is supposed to represent. Could you please point out where am I wrong in my thought process?

**Comment by argowang** [comment]
> Note has the signature field because the intention is that the BuildProvenance in the Occurrence would be signed and the Build Note created with this signature.

Hi @aysylu, I am having same issue as @ba32107 to understand the signature. From the quoted sentences, it sounds like we need to know the provenance when creating the note? Also, from what it appears, all occurrences under the same note need to share the same provenance. Otherwise, the signature validation would fail.

I am sure @ba32107 and I misunderstood some concepts here. Would it be possible for you to provide a concrete example for us to understand?

Thanks.

**Comment by ba32107** [comment]
Hi @aysylu, did you have a chance to look into this question please?

**Comment by argowang** [comment]
With another look, it looks like there is a one-to-one mapping between the build note and build occurrence. 

@aysylu Correct me if I'm wrong. But based on the comments in the proto file, I believe here is how it works:
1. In one Jenkins job, we collect the meta-data and construct `provenance` and  `provenance_byte`
2. In the job, we signed the `provenance_byte` with a private key. Then we create a `build note` with the signature and the info about the `public key`
3. After the note is created, we create a `build occurrence` and point it to the `build note` we created in step 2.

When verifying, we use the signature stored in the `build note` and the corresponding public key and the `provenance_byte` to ensure integrity.

Based on my understanding, each Jenkins job would have a separate `build note`. Each `build note` would only have **one** `build occurrence`.

**Comment by ba32107** [comment]
> Based on my understanding, each Jenkins job would have a separate build note. If there are multiple images created in the job, each image would have a build occurrence. Since all images were built in the same job, they would have the same provenance.

I believe this is not the correct assumption. This would only work if every occurrence would have the exact same build provenance. First, this is not possible because the provenance not only contains the source information, but the exact build commands as well. So if a job produces multiple output images, they would need to have different build provenances (because the only way to produce different images is with different build commands). Second, if all build occurrences are supposed to have the exact same provenance, then there is no point in having multiple occurrences. There is no extra information the individual occurrences hold.

I am fairly sure that the `v1beta1` API spec is simply wrong. I don't see how it makes sense. The `v1` spec does make sense (because there is no signature in the Note), but it loses on the benefit of cryptographic signatures. Hopefully it's going to be added (to the `BuildOccurrence`, where it belongs).

I may be wrong though, would love to have some clarification from the maintainers.

**Comment by wkozlik** [comment]
Contributions to improve this are welcome.
